scaaml/io/resume_kti.py

```diff
@@ -0,0 +1,156 @@
+"""Resumeable iterator for key-text pairs with autosaves.
+
+Autosaves after shard length iterations.
+
+Typical usage example:
+
+  # Save all key-text pairs ahead before the experiment starts.
+  # If kt_filename exists this does nothing.
+  create_resume_kti(keys, texts, shard_length=shard_length,
+                    kt_filename="key_text_pairs.txt",
+                    progress_filename="progress_pairs.txt")
+
+  # Load the savepoint (all the pairs + the index where to continue) at the
+  # start or resume of the experiment.
+  resume_kti = ResumeKTI(kt_filename="key_text_pairs.txt",
+                         progress_filename="progress_pairs.txt"
+
+  # for key, text in tqdm(resume_kti):  # Loop with a progress-bar.
+  for key, text in resume_kti:
+      # Make sure to capture the trace.
+      while True:
+          # Try to capture trace.
+          if trace:
+              break
+
+File format:
+
+  Text-key pairs file is a binary file with numpy scalar and two arrays.  The
+  contents are shard length, keys, texts. The array of keys and texts are
+  interpreted in the way that (keys[i], texts[i]) is a single key-text pair for
+  any i in range(len(keys)), thus both have the same length. This is done in
+  order to allow different data type of key and text (for instance AES256 where
+  key is 256 bits and text is 128 bits).
+
+  Progress file is a text file with a single integer -- the number of traces
+  captured so far. The number in it is a multiple of the shard length. This file
+  is automatically updated after shard length iterations over ResumeKTI.
+"""
+
+import numpy as np
+
+
+def create_resume_kti(keys: np.ndarray,
```

Comment:
Is there a case when we would call `create_resume_kti()` without almost immediately calling `ResumeKTI.__init__()` after?
If not I suggest to call the constructor at the end of `create_resume_kti()`.

Reply:
Done.